h5py/_conv.pyx

```diff
@@ -153,113 +134,98 @@ cdef int conv_vlen2str(void* ipt, void* opt, void* bkg, void* priv) except -1:
+    cdef:
+        PyObject** buf_obj = <PyObject**>ipt
+        char** buf_cstring = <char**>opt
+        conv_size_t* sizes = <conv_size_t*>priv
+        char* temp_string = NULL
+        size_t temp_string_len = 0  # Not including null term
+        PyObject* buf_obj0
+        char* buf_cstring0
+        object temp_object
 
-    cdef PyObject** buf_obj = <PyObject**>ipt
-    cdef char** buf_cstring = <char**>opt
-    cdef conv_size_t* sizes = <conv_size_t*>priv
-
-    cdef PyObject* temp_object = NULL
-    cdef PyObject* temp_encoded = NULL
+    buf_obj0 = buf_obj[0]
+    if buf_obj0 == NULL:
+        temp_object = None
+    else:
+        temp_object = <object> buf_obj0
 
-    cdef char* temp_string = NULL
-    cdef size_t temp_string_len = 0  # Not including null term
+    if temp_object is None:
+        temp_string = ""
+        temp_string_len = 0
+    else:
 
-    cdef PyObject* buf_obj0
-    cdef char* buf_cstring0
+        if isinstance(temp_object, unicode):
+            temp_object = temp_object.encode('utf-8')
 
-    memcpy(&buf_obj0, buf_obj, sizeof(buf_obj0))
+        elif not isinstance(temp_object, bytes):
+            # There is not test on this !
+            if sizes.cset == H5T_CSET_ASCII:
+                temp_object = bytes(temp_object)
+            elif sizes.cset == H5T_CSET_UTF8:
+                temp_object = str(temp_object)
+            else:
+                raise TypeError("Unrecognized dataset encoding")
 
-    try:
-        if buf_obj0 == NULL or buf_obj0 == Py_None:
-            temp_string = ""
-            temp_string_len = 0
-        else:
-            if PyBytes_CheckExact(buf_obj0):
-
-                # Input is a byte string.  If we're using CSET_UTF8, make sure
-                # it's valid UTF-8.  Otherwise just store it.
-                temp_object = buf_obj0
-                Py_INCREF(temp_object)
-                if sizes.cset == H5T_CSET_UTF8:
-                    try:
-                        pass # disabled for Python 3 compatibility
-                        #temp_encoded = PyString_AsDecodedObject(temp_object, "utf8", NULL)
-                    except:
-                        raise ValueError("Byte string is not valid utf-8 and can't be stored in a utf-8 dataset")
-                temp_string = PyBytes_AsString(temp_object)
-                temp_string_len = PyBytes_Size(temp_object)
-
-            # We are given a Unicode object.  Encode it to utf-8 regardless of
-            # the HDF5 character set.
-            elif PyUnicode_CheckExact(buf_obj0):
-                temp_object = buf_obj0
-                Py_INCREF(temp_object)
-                temp_encoded = PyUnicode_AsUTF8String(temp_object)
-                temp_string = PyBytes_AsString(temp_encoded)
-                temp_string_len = PyBytes_Size(temp_encoded)
+        if sizes.cset == H5T_CSET_UTF8:
+            try:
+                temp_object.decode('utf-8')
```

Comment:
Are you speaking of h5py 3.0 ... because all the string handling is already performed in python3 syntax

Reply:
Yes, I meant h5py 3.0. See #1338 for what's planned. For this PR, though, let's keep things working the way they do in h5py 2.x, otherwise it will be much harder to review.

Reply:
This `try:.. except` part can be removed, it was there to replace a test that was comment 8 years ago (https://github.com/h5py/h5py/commit/db045a12ded84327a48ea062e1c6b0e59ce20998)